### DISCUSSION TITLE
Expose Hono server for middleware and endpoints?

### DISCUSSION THREAD

**Discussion by Callumk7** [wakujs/waku#847] [General]
Good morning! 
Firstly, thanks for this project. It is really fascinating to see how you have put it together. 

I was wondering if you have plans for exposing a mechanism for adding api routes and middleware via the Hono server that is being used within the project? If it is not too early to share, how might this API look?

Cheers,
Callum

**Comment by dai-shi** [comment]
Hi! Thanks for asking.
We already have (undocumented) middleware capability, but it doesn't use Hono API.
And, we plan to add api endpoint capability in v0.22.
Would you share your thoughts about the use case? Does it require Hono API?

**Comment by Callumk7** [comment]
Great! for middleware, that is great to hear. There is no specific reason I have for middleware to utilise hono, that was just an assumption I made. I look forward to seeing it when it is ready to be shown.

Regarding use-cases, I tend to find that there hits a point in my Remix projects where I move my API functionality to a separate hono/express app as file-based routing can become quite messy quite quickly. I use Remix as a backend-for-frontend to handle SSR, and make calls to my backend server which handles the db. Remix builds to a request handler, and there are solutions/tempalates that exist that essentially slap it onto a hono/express server as a wildcard endpoint, meaning you can utilise loaders/actions whilst also building out a more 'express-style' application within the same project for deployment. This is actually also the only way you can use middleware with remix, until they release their own implementation in the not-too-distant future (I think).

Seeing as hono is already here, I was just musing that this might be a lot tidier with waku in theory, if one were to want to achieve something similar. I am aware though that I am focusing on the fact that you are already using hono though, rather than taking a more holistic view of the problem and the solutions you might be looking at instead.

**Reply by dai-shi** [reply]
Thanks for sharing your ideas. That sounds like very close to the reason we plan to add API endpoint support for the next minor.
Having that said, I'm eager to learn use cases from someone who has some experience, and also learn what would the limitation with server actions.

If you can help me, I would like you to create some smaller examples:
- one Waku app that uses server actions.
- one Remix app that does mostly the same.
- one non-working Waku app code that you would expect with API endpoint support.

Not in a hurry though.
